README.md
### nRF51duino

This code supports nRF51822 and nRF41422 chips from Nordic Semiconductor. At the moment no
Softdevice functionality is supported out of the box. A sketch cannot be compiled without
Nordic SDK. Compliling a sketch without SDK and a native Radio Library (nRF24 compatible)
is planned. Feel free to add Softdevice or native BLE support so BLE or ANT+ can be used
without Nordic nRF51 SDK.

Please keep in mind that using Nordic SDK is not open source friendly at the moment. All
example code and a lot of header files doesn't allow redistribution of source code.

To start you need an nRF51 development board with an JLink Adapter. You need to buy an offical
development kit. If you plan to use more than one board choose an Development Kit with an external
JLink adapter. If you want to use ANT+ protocol choose an nRF51422 or nRF51922 kit.

Whats not working at the moment
 - Flashing Software under Windows
 - A lot of untested functionality

Tested at the moment:
 - Startcode
 - RTC
 - DigitalWrite
 - Delay
 - Serial interface

I Need help to:
 - remove linker warnings
 - write libraries for integrated chip hardware (Random, AES, CRC, Radio, Timer, RTC0, ...)
 - port my flash script to windows
 - test my flash script on OSX
 - remove Nordic SDK dependencies
 ...

Not planned by me, but a good idea making this software better:
 - support SoftDevices without Nordic SDK (https://github.com/mrquincle/bluenet) or Porting a
   complete BLE stack http://code.google.com/p/btstack/
 - A Jlink indipended Bootloader

This is a fork of RFduino with broken compatibility because there are a lot of non open source
files in RFduino. I have broken command compatibility to get more indipended and make it easy
recreate core functionality. Thanks RFduino for your preliminary work.

WARNING: Please don't connect more then one board at the same thime when flashing.

### Installation

* Buy an Nordic NRF51 HDK, Download SDK and JLink Software

* Download Arduino 1.5 here: [http://arduino.cc](http://arduino.cc/en/Main/Software)  
  (on osx remember to open arduino first to make gatekeeper perform its magic)  

* Copy the nRF51duino directory from this repository in Arduino  
  (on Windows, C:\arduino-1.5.4\hardware\arduino)  
  (on OSX, /Applications/Arduino.app/Contents/Resources/Java/hardware/arduino)  
  or "git clone https://github.com/d00616/nRF51duino" in the directory indicated

* Copy Nordic NRF51 SDK to hardware/arduino/nRF51duino/system/nRF51-SDK

* Install SEGGER JLINK Software

Your ready to go!

### Getting Started

* Select a nRF51 Board from the Tools/Board menu

* Select an example from the Files/Examples/RFduinoNonBLE directory

* Select Upload to compile, upload and execute the sketch

### Define you own board layout

nRF51duino comes with predifinions for an PCA10000 Board. You can create your
own "custom_board.h" file in your Sketch to override settings.


CHANGELOG
CHANGELOG

v 3.0
   - Fork project to nRF51duino
   - Remove RFduino specific files
   - New, open source core code
   - rename function calls and files
   - Linux flash script

===RFduino====

v 2.2
  - Support for nonconnectable advertisement
  - Corrected temperature comments (thanks DasBasti).

v 2.1.2
  - Support for MIDI BaudRate of 31250 (thanks tolson)
  - Fix so Servo can be moved from TIMER2 to TIMER1 for use with RFduinoGZLL

v 2.1.1
  - Fix an issue with RFduinoBLE.begin() not starting in an isolated case.
  - Fix to case of include directory in boards.txt (thanks jrobeson).

v 2.1
  - Fixes required to support Arduino 1.5.7
    (NOTE: Arduino 1.5.7 or higher is now required for RFduino 2.1 or above).
  - Updated gcc compiler now generates smaller code sizes
  - Sketch size information now includes RAM usage.

v 2.0.3
  - Added printf() support to Serial (see RFduinoNonBLE/Printf sketch)
  - Fixed issue with Serial.flush()/Serial.end() when switching TX/RX pins

v 2.0.2
  - Added connection interval update support
  - Added missing copyright notices to sketches

v 2.0.1
  - RFduinoGZLL enhancement for channel selection policy
  - allow build.bat "run as administrator"

v 2.0
  - 128-bit custom UUID enhancement
  - raw advertisement packet access
  - RFduinoBLE.begin() now truncates (instead of failing) if the advertisement
    packet length is exceeded.  No more counting deviceName/advertisementData.
    
v 1.9
  - i2c enhancements
  - UART TX buffering
  - Changed examples comment to indicate that Advertisement length and
    deviceName length must be <= 15 bytes
  - Added header to WString.h to so F() always available

v 1.8.2
  - Added support for modifying the iBeacon measured power level at 1 meter
  - Fix for analogWrite and RFduinoGZLL (thanks tolson)

v 1.8.1
  - Fixed conflict between Analog to Digital Conversion and UART
  - Cleaned up SPI library

v 1.8
  - iBeacon Enhancement

v 1.7.2
  - Enhancement to PPI channel selection to eliminate potential conflicts
  - Added the ability to select I2C speed to Wire Library (100, 250 or 400 kbps)

v 1.7.1
  - Fix for the UART stops receiving bytes after some time with BLE enabled

v 1.7
  - Added RFduino to RFduino wireless communication support
  - Added pgmspace.h / PROGMEM support
  - Added __RFduino__ constant
  - Provided a 32-bit RFDLoader for 32-bit OSX support
  - Cleaned up startup code and dependencies

v 1.6.1
  - Fixed issue with not being able to call RFduinoBLE.end() / 
    RFduino_systemReset() from pinWakeCallback
  - corrected return types from uint32_t to void for rfduino_ functions
  - Corrected maximum sketch size.  The last 4 flash pages are reserved,

v 1.6
  - Added getDeviceId() to return the unique id for each device
    (see DeviceInfo sketch)

v 1.5.1
  - Added support for OSX (contributed by nitsky)
  - RFDLoader now supports 32-bit and 64-bit osx snow leopard through mavericks
  - fix for i2c not working when BLE stack is active

v 1.5
  - Added support for Flash Memory.  See the RFduinoNonBLE flash example
    sketches for use.  More documentation is also available in
    cores/arduino/Memory.h
  - Added a ProximityMovingAverage sketch for investigating RSSI to distance
    calculations
  - Fix for iOS SDK 7.0 issue with the disconnect characteristic

v 1.4
  - Integrated RFDLoader

v 1.3.2
  - Fixed issue with the send characteristic always sending 20 bytes
  - Increased send characteristic payload size from 12 to the maximum of 20
    bytes
  - Added sketch and iPhone demonstrating lossless bulk data transfer

v 1.3.1
  - Fixed issue with on_rssi() only working for the first connect

v 1.3.0
  - Added support for radio active notification to support sketches with
    critical timing requirements.

v 1.2.0
  - Restructured the RFduino repository to allow a git clone directly into the
    Arduino directory structure.

v 1.1.2
  - Corrected the name of a missing extern.
  - UART_DEFAULT_RX_PIN and UART_DEFAULT_TX_PIN and multiply defined (and
    conflicting).  Removed the erroneous definitions.

v 1.1.1
  - Added support for OSX Snow Leopard.

v 1.1.0
  - Support for OSX.
  - equivalent to RFduino305.zip.

v 1.0.0
  - equivalent to RFduino289.zip.


variants/nRF51/variant.h
/*
 Copyright (c) 2014 nRF51duino.  All right reserved.

 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 2.1 of the License, or (at your option) any later version.

 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 See the GNU Lesser General Public License for more details.

 You should have received a copy of the GNU Lesser General Public
 License along with this library; if not, write to the Free Software
 Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

/*
 Copyright (c) 2013 OpenSourceRF.com.  All right reserved.

 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 2.1 of the License, or (at your option) any later version.

 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 See the GNU Lesser General Public License for more details.

 You should have received a copy of the GNU Lesser General Public
 License along with this library; if not, write to the Free Software
 Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

/*
  Copyright (c) 2011 Arduino.  All right reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  See the GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

/*
 * Here should go some sort of liscense maybe? This file is a modified
 * arduino source file.  It defines the board variant being used
*/

#ifndef _VARIANT_NRF51DUINO_
#define _VARIANT_NRF51DUINO_

/*----------------------------------------------------------------------------
 *        Headers
 *----------------------------------------------------------------------------*/

#include "Arduino.h"

#include "custom_board.h"

#ifdef __cplusplus
#include "UARTClass.h"
#endif

#ifdef __cplusplus
extern "C"{
#endif // __cplusplus

/**
 * Libc porting layers
 */
#if defined (  __GNUC__  ) /* GCC CS3 */
#    include <syscalls.h> /** RedHat Newlib minimal stub */
#endif

/*----------------------------------------------------------------------------
 *        Definitions
 *----------------------------------------------------------------------------*/

/*----------------------------------------------------------------------------*/

#define NRF51DUINO

/** Name of the board */
#define VARIANT_NAME        "nRF51duino"

/** Master clock frequency  (used in TWI and DAC) */
#define VARIANT_MCK         16000000UL

/*----------------------------------------------------------------------------
 *        Pins
 *----------------------------------------------------------------------------*/

// Number of pins
#define PINS_COUNT           (7u)

/* Pins descriptions to be instantiated into variant.cpp */
extern uint8_t PPI_Channels_Occupied[PINS_COUNT][2];
extern uint8_t GPIOTE_Channels_Occupied[PINS_COUNT];
extern uint8_t Timer1_Compare_Unit_Occupied_by_Pin[3]; // Contains Pin number which occupied appropriate CC register (0, 1 or 2) of Timer 1 for PWM purposes
extern uint8_t Timer2_Compare_Unit_Occupied_by_Pin[3]; // Contains Pin number which occupied appropriate CC register (0, 1 or 2) of Timer 2 for PWM purposes
extern uint8_t Pin_Occupied_for_PWM[PINS_COUNT]; // Determines if particular pin is occupied for PWM output

/*
 * SPI Interfaces
 */
#define SPI_INTERFACES_COUNT 2

#define SPI_INTERFACE        NRF_SPI0
#define PIN_SPI_SS           (6u)
#define PIN_SPI_MOSI         (5u)
#define PIN_SPI_MISO         (3u)
#define PIN_SPI_SCK          (4u)

#define SS    PIN_SPI_SS
#define MOSI  PIN_SPI_MOSI
#define MISO  PIN_SPI_MISO
#define SCK   PIN_SPI_SCK

/*
 * Wire Interfaces
 */
#define PIN_WIRE_SDA         (6u)
#define PIN_WIRE_SCL         (5u)
#define WIRE_INTERFACE       NRF_TWI1

/*
 * Analog pins
 */
#define ADC_RESOLUTION        10

/*
 * UART/USART Interfaces
 */
#define UART_DEFAULT_RX_PIN		0
#define UART_DEFAULT_TX_PIN		1

/*
typedef enum
{
  UART0_State_NotStarted,
  UART0_State_BeforeFirstTX,
  UART0_State_AfterFirstTX
}  UART0_States;

extern UART0_States UART0_State;
int UART0_BaudRate();
void UART0_Start( int dwBaudRate, uint8_t rx_pin, uint8_t tx_pin, uint8_t rts_pin, uint8_t cts_pin );
void UART0_Stop();
void UART0_FlushTX();
void UART0_TX( const uint8_t uc_data );
void UART0_TXD( const uint8_t uc_data );

__attribute__(( always_inline ))
static int __INLINE UART0_RXReady()
{
  return NRF_UART0->EVENTS_RXDRDY;
}

__attribute__(( always_inline ))
static int __INLINE UART0_TXReady()
{
  return NRF_UART0->EVENTS_TXDRDY;
}

__attribute__(( always_inline ))
static uint8_t __INLINE UART0_RXData()
{
  return NRF_UART0->RXD;
}

void UART0_RXReset();
int UART0_RXErrorReset();
uint8_t UART0_RX();

void UART0_TXReset();
*/

/*
 * PWM
 */
#define PWM_MAX_DUTY_CYCLE    255
#define PWM_MIN_DUTY_CYCLE    0
#define PWM_RESOLUTION        16

/*
 * Miscellaneous
 */

#include "Memory.h"

// unique id for each unit
uint32_t getDeviceIdLow();
uint32_t getDeviceIdHigh();
uint64_t getDeviceId();

extern bool override_uart_limit;

#define UART_BAUDRATE_BAUDRATE_Baud31250 (0x00800000UL) /*!< 31250 baud.(MIDI)  */

#ifdef __cplusplus
}
#endif

/*----------------------------------------------------------------------------
 *        Arduino objects - C++ only
 *----------------------------------------------------------------------------*/

#ifdef __cplusplus

extern UARTClass Serial;

#endif

#endif /* _VARIANT_NRF51DUINO_ */
